Experiment: Viper integration

rvanasa opened the pull request:
VS Code support for https://github.com/dfinity/motoko/pull/3477.

- This implementation uses `moc.js` (rather than the OCaml LSP) to generate Viper files. 
  - This makes it possible for anyone to try the demo without requiring a modified `moc` or `dfx` installation. 
  - See below for steps to quickly rebuild the extension. 
- Activate code translation by adding `// @verify` at the top of a Motoko file. 

Example (`Main.mo`):

```motoko
// @viper

actor {
  var claimed = false;
  var count = 0 : Int;

  assert claimed and not (-1 == -1) and (-42 == -42) or true;
  assert count > 0;

  public shared func claim() : async () { };
};
```

Generated file (`Main.vpr`):
```viper
method __init__($Self: Ref)  
  ensures invariant_7($Self)
  ensures invariant_8($Self)
   { 
     ($Self).claimed := false
     ($Self).count := 0 
   }
field claimed: Bool
field count: Int
define invariant_7(self) (((((self).claimed && (!(-1 == -1))) && (-42 == -42)) || true))
define invariant_8(self) (((self).count > 0))
method claim($Self: Ref) 
  requires invariant_7($Self)
  requires invariant_8($Self)
  ensures invariant_7($Self)
  ensures invariant_8($Self)  { 
                                 
                              }
```

Recommended editor workflow using split-screen tabs (updates as you type):

<img width="1728" alt="Screen Shot 2022-10-17 at 6 13 04 PM" src="https://user-images.githubusercontent.com/522097/196306593-8690526d-1022-43f8-a15f-e547529090e1.png">


Steps to build the extension:
- Ensure `motoko` and `vscode-motoko` are in the same parent directory
- Switch to the `ryan/viper` branch in the Motoko compiler repository
- In your terminal, run `cd vscode-motoko` and then `npm run package` (this will rebuild `moc.js`)
- Right-click the generated `/vscode-motoko-*-viper.vsix` file and select "Install extension VSIX"

Progress:
- [x] Generate Viper files on code change
- [x] Show Motoko diagnostics from `Pipeline.viper_files`
- [x] Show Viper diagnostics in Motoko via source map

rvanasa committed:
Viper integration progress

rvanasa committed:
Progress

rvanasa committed:
Merge branch 'master' into viper

rvanasa committed:
Add Viper as an extension dependency

rvanasa committed:
Merge branch 'viper' of https://github.com/dfinity/vscode-motoko into viper

rvanasa committed:
Add '-viper' suffix to extension version

rvanasa committed:
Optimize and clarify type checking

rvanasa committed:
Add generated 'moc.js' file to gitignore

rvanasa committed:
Add boilerplate for Viper LSP errors + source mapping

rvanasa committed:
Implement verification RPC calls

rvanasa committed:
Checkpoint: receiving diagnostics from StateChange notifications

rvanasa committed:
Add basic Motoko diagnostics from Viper LSP

rvanasa committed:
Add 'View in context' link from Motoko -> Viper diagnostics

rvanasa committed:
Adjust diagnostic source messages

rvanasa committed:
Update build script to account for z3 dependency

rvanasa committed:
Misc

rvanasa committed:
Add 'implies' and 'invariant' to syntax highlighting

rvanasa commented on package.json:
```diff
@@ -155,9 +155,14 @@
-        "compile": "rimraf ./out && tsc -p .",
+        "vscode:prepublish": "npm run generate && npm run compile",
+        "generate": "npm run download && cd ../motoko && nix-shell --command 'make -C src moc.js' && cp -vL src/moc.js ../vscode-motoko/src/generated",
+        "download": "curl -L -o src/generated/viperserver.jar https://github.com/viperproject/viperserver/releases/download/v-2022-10-18-0728/viperserver.jar -C -",
```
Good idea; now using the Nix environment for the stand-in portable logic.

rvanasa commented:
Progress:

- Implemented error reporting in Motoko files from a sandboxed Viper LSP
- Added syntax highlighting for the `invariant` and `implies` keywords
- Included a "view in context" button to jump from Motoko errors to the corresponding Viper source location

<img width="1728" alt="Screen Shot 2022-10-20 at 10 15 48 PM" src="https://user-images.githubusercontent.com/522097/197110607-acf0097a-a5a2-42aa-a790-b8a9ec4f234b.png">

In this screenshot, the `// @viper` comment is being used as a default error location. The TS language server is ready to display errors at the correct locations once the lookup function from `moc.js` is working as expected (https://github.com/dfinity/motoko/pull/3501).

aterga commented:
While installing the extension from this branch, I've experienced the following minor hick-ups:

1. Needed to run `npm install run-a --dev` (maybe repeating `npm install` would have sufficed)
2. Needed to run `src/generated` (otherwise, I was getting 

```
'src/moc.js' -> '../vscode-motoko/src/generated/moc.js'
cp: cannot create regular file '../vscode-motoko/src/generated/moc.js': No such file or directory
```

The rest of the instructions worked out nicely!

PS. I ran `npm run package` from `motoko` repo's nix-shell.

aterga commented:
Also, one probably needs `code --force --install-extension `pwd`/vscode-motoko-0.5.3-viper.vsix`

crusso commented:
> 2. Needed to run `src/generated` (otherwise, I was getting

An you mean create (not run) src/generated.

rvanasa committed:
Fix initial setup process in package.json

rvanasa commented:
> While installing the extension from this branch, I've experienced the following minor hick-ups:

Fixed; thanks!

rvanasa committed:
Fix

rvanasa committed:
// @viper -> // @verify

rvanasa committed:
Only insert Viper code into '*.vpr' files

rvanasa committed:
Rewrite verification error messages

rvanasa committed:
Improve unknown range handling

rvanasa committed:
Hide unused Viper LS message types

rvanasa committed:
Add provisional success diagnostic

aterga commented on src/server/viper.ts:
```diff
@@ -0,0 +1,321 @@
+        const [, method] = match;
+        return `Canister invariant violated by method '${method}'`;
+    }
+    return message;
```
```suggestion
    if (message.startsWith('Exhale might fail. Assertion ') {
        return 'Canister invariant violated by async block'
    }
    return message;
```

ggreif commented on package.json:
```diff
@@ -155,9 +155,15 @@
         "vsce": "^2.11.0"
     },
+    "extensionDependencies": [
+        "viper-admin.viper"
```
He already uses `$VIPER_SERVER`! But the ideal thing would be to frob the port number form the `viper-ide` panel, and connect to the already running server.

rvanasa replied:
This would be ideal. The main challenge here is that VS Code loads extensions based on [activation events](https://vscode-docs.readthedocs.io/en/stable/extensionAPI/activation-events/), so it's currently required to manually open a `*.vpr` file to activate the Viper LS. 

Update: I just found a way to programmatically activate the Viper extension using the [extensions API](https://code.visualstudio.com/api/references/vscode-api#extensions), so now we just need a way to reliably retrieve the port number.

ggreif commented on src/server/viper.ts:
```diff
@@ -0,0 +1,271 @@
+                            );
+                            await connection.sendNotification(
+                                new rpc.NotificationType(
+                                    'textDocument/didSave',
```
IIRC, this is not needed. Just be sure to set `{ manuallyTriggered : true }`.

rvanasa replied:
Removed, and everything still works! Also updated `manuallyTriggered : true`.

rvanasa commented on src/server/viper.ts:
```diff
@@ -0,0 +1,271 @@
+                                    diagnostic.range,
+                                ) || {
+                                    start: { line: 0, character: 0 },
+                                    end: { line: 0, character: 100 }, // Highlight the `// @viper` comment by default
```
VS Code automatically limits the maximum character length, so this works as intended with Error Lens. 

I'm planning to replace this with a checkmark decoration, so this is mostly a stand-in feature to confirm that we are receiving the "success" message from the Viper LS. 

<img width="672" alt="Screen Shot 2022-10-21 at 9 25 44 PM" src="https://user-images.githubusercontent.com/522097/197317092-06ea5602-8f7d-4e0f-af2e-c76035617570.png">

rvanasa commented on src/server/viper.ts:
```diff
@@ -0,0 +1,271 @@
+                console.log(msg);
+            } else {
+                console.log(`[${msgType}]`);
+            }
```
This is mostly debug output from the Viper LS (such as `ast_construction_success`, `configuration_confirmation`, `copyright_report`, etc.), so we can probably ignore these messages entirely now that we are properly receiving messages from the language server.

rvanasa committed:
Change error message (lifting error information from Viper)

Co-authored-by: Arshavir Ter-Gabrielyan <arshavir.ter.gabrielyan@dfinity.org>

rvanasa commented on src/server/viper.ts:
```diff
@@ -0,0 +1,271 @@
+        }
+        return result.diagnostics;
+    } catch (err) {
+        console.error(`Error while translating to Viper: ${err}`);
```
Good catch--thanks!

rvanasa commented on src/server/viper.ts:
```diff
@@ -0,0 +1,271 @@
+
+            // Debounce verification
+            clearTimeout(verifyTimeout);
+            verifyTimeout = setTimeout(() => {
```
It was necessary to add a debounce (using the confusingly named JS `setTimeout` API) to avoid hitting a limit to the number of active verifications in the Viper LS. This logic waits to start verification until 500ms after a user finishes typing (rather than on every keystroke), which was more than enough to resolve this issue.

rvanasa committed:
Misc fixes from code review

rvanasa committed:
Merge branch 'viper' of https://github.com/dfinity/vscode-motoko into viper

ggreif commented on src/server/viper.ts:
```diff
@@ -0,0 +1,324 @@
+                            ? viperDiagnostics
+                            : [
+                                  {
+                                      message: 'Verification succeeded',
```
```suggestion
                                      message: '✅ Verification succeeded',
```

crusso commented:
Great stuff! Thanks again Ryan!

rvanasa committed:
Run 'npm audit fix'

rvanasa committed:
Fix verification success corner case

ggreif committed:
tweaks (#80)

Two minor changes
- fixed "The latest version of `vsce` is 2.13.0 and you have 2.11.0" warning
- save space by hard-linking contents of `src/generated` to `out/generated`.

The former looked like low-hanging fruit.
The latter is important for me, as my Mac has a 256 GB disk that is almost full :-)
```
[nix-shell:~/vscode-motoko]$ ls -l out/generated
total 124668
-r--r--r-- 2 ggreif staff 22063062 Oct 23 02:31 moc.js
-r--r--r-- 2 ggreif staff 81300825 Oct 23 02:31 viperserver.jar
-r-xr-xr-x 2 ggreif staff 24292476 Oct 23 02:31 z3
```

The `package-lock.json` changes were obtained by running `npm audit fix`.

ggreif commented on src/server/viper.ts:
```diff
@@ -0,0 +1,331 @@
+
+    connection.onRequest(new rpc.RequestType('GetViperFileEndings'), () => {
+        return {
+            fileEndings: ['*.vpr'],
```
```suggestion
            fileEndings: ['*.mo.vpr'],
```

ggreif commented:
I am findig
```
$ grep 'ViperServer online at' /tmp/.vscode/viper.log 
00:03.564 . Server: [ViperServer]: ViperServer online at http://localhost:50002
```

rvanasa committed:
Use '*.mo.vpr' file endings

Co-authored-by: Gabor Greif <gabor@dfinity.org>

rvanasa commented:
> Pre-release is described here: https://code.visualstudio.com/api/working-with-extensions/publishing-extension#prerelease-extensions

It's possible to switch to 4.x by selecting "Viper" in the Extensions panel and then clicking "Switch to Pre-Release Version" (screenshot for clarity):

<img width="1728" alt="Screen Shot 2022-10-23 at 12 11 05 PM" src="https://user-images.githubusercontent.com/522097/197408566-54ce53b6-459d-490b-96e8-a739866302d9.png">

I've encountered a variety of issues while using the pre-release version, but it should be possible to work around or patch them as needed.

rvanasa committed:
Improve verification success diagnostic

rvanasa committed:
Merge branch 'viper' of https://github.com/dfinity/vscode-motoko into viper

rvanasa committed:
Improve verification success detection

rvanasa committed:
Switch RPC 'Verify' request to notification

rvanasa committed:
Merge branch 'master' of https://github.com/dfinity/vscode-motoko into viper

rvanasa committed:
Add icon for Motoko files

rvanasa committed:
Merge branch 'master' into viper

rvanasa committed:
Merge branch 'master' into viper

ggreif committed:
remove premature invalidation

The reason is …. cache invalidation!
There is a curious order of events:
> compileViper (mocViperCache.set) /Users/ggreif/motoko/test/viper/claim.mo
> writeVirtual (invalidateViper) /Users/ggreif/motoko/test/viper/claim.mo
the latter clearing the cache :zany_face:. I have commented out the invalidation from `writeVirtual` on `viper` branch

rvanasa committed:
Disable workspace type-checking

ggreif committed:
LSP: wait till port is up and parse it (#86)

This captures the LSP's output and waits until it announces its listening port. Then the connection is set up towards that port.

- Avoids premature connect when the listening port is not ready yet. (This happened routinely on my Mac.)
- Also avoids the fixing of a port number for LSP.
- Also eliminates Viper-related junk from the distribution (reducing size) and reuses the artefacts that come with Viper itself.

_CAVEAT_: You'll need to install the "pre-release" Viper extension from the marketplace and (suggested) switch the tool download to `Nightly`.

Open issue:
- locate the `java` binary (there is an npm for that!)

Co-authored-by: rvanasa <ryan.vandersmith@dfinity.org>

rvanasa committed:
Look up Java home directory using viper-ide logic (#91)

rvanasa committed:
Merge branch 'master' of https://github.com/dfinity/vscode-motoko into viper

rvanasa committed:
Update marketplace details in package.json

rvanasa committed:
Update readme for VS Marketplace

aterga committed:
Add TM for Formal Motoko  (#87)

* Add color theme supporting static annotations, e.g., assert:invariant

* color specs in coral

* Support highlighting of missing static annotations in syntaxes/Major.tmLanguage

Co-authored-by: Ryan Vandersmith <ryan.vandersmith@dfinity.org>

rvanasa committed:
Merge branch 'viper' of https://github.com/dfinity/vscode-motoko into viper

rvanasa committed:
Remove custom 'Formal Motoko' theme in favor of using light/dark themes provided by Viper

rvanasa committed:
Update package-lock.json

rvanasa committed:
Set 'MOC_UNLOCK_VERIFICATION' environment variable for moc.js

ggreif committed:
some version bumps (#97)

ggreif committed:
chore: set up a `motoko` submodule (#98)

* set up a `motoko` submodule
* update instructions
* Update README.md
* Update .vscodeignore
* add a missing item

Co-authored-by: rvanasa <ryan.vandersmith@dfinity.org>

rvanasa committed:
Replace tsc with esbuild bundler

rvanasa committed:
Fix missing Wasm file for 'prettier-plugin-motoko'

rvanasa committed:
Fix .vscodeignore

ggreif committed:
simplify instructions

rvanasa committed:
Merge branch 'bundle-esbuild' into viper

rvanasa committed:
Add missing dev-dependency

rvanasa committed:
Add 'node_modules' back into .vscodeignore

ggreif committed:
fix

rvanasa committed:
Replace 'mkdir -p' with 'mkdirp' for Windows build support

rvanasa committed:
Merge branch 'master' of https://github.com/dfinity/vscode-motoko into viper

rvanasa committed:
Update readme

rvanasa committed:
Add incompatibility prompt

rvanasa committed:
Use custom syntax highlighting provided by Viper

rvanasa committed:
Add 'Background' section to readme

rvanasa committed:
Temporarily rename extension to 'Formal Motoko'

rvanasa committed:
Track debug launch configuration

aterga committed:
First (#102)

rvanasa committed:
Rename 'Formal Motoko' -> 'Motoko-san'

rvanasa committed:
v0.0.2

ggreif committed:
chore: Viper is on Motoko `master` (#103)

the `viper` branch in https://github.com/dfinity/motoko is merged (and thus deleted)

rvanasa committed:
Merge branch 'master' into viper

rvanasa committed:
Misc. fixes

rvanasa committed:
Merge remote-tracking branch 'origin/master' into viper

rvanasa committed:
Improve diagnostic behavior when opening/closing files

aterga commented on src/server/viper.ts:
```diff
@@ -0,0 +1,345 @@
+    );
+    if (match) {
+        const [, method] = match;
+        return `Canister invariant violated by method '${method}'`;
```
We need to make this slightly more general, as we now get the following possibility:

<img width="509" alt="Screenshot 2023-01-04 at 20 07 18" src="https://user-images.githubusercontent.com/886924/210630710-8d4f2211-2192-424f-a728-bf10e8288308.png">

Here, it's a postcondition violation, but we cannot easily distinguish between violations of "regular" postconditions vs. those representing a canister invariant check. 

```suggestion
        return `Canister specification violated by method '${method}'`;
```

aterga commented on syntaxes/Major.tmLanguage:
```diff
@@ -737,7 +745,7 @@
 					<string>statement keyword</string>
 					<key>match</key>
-					<string>\b(assert|break|case|continue|default|debug|debug_show|else|if|ignore|in|for|label|null|return|switch|while|loop|try|throw|catch|do|to_candid|from_candid|with)\b</string>
+					<string>\b(assert|break|case|continue|default|debug|debug_show|else|if|ignore|in|for|label|null|return|switch|while|loop|try|throw|catch|do|to_candid|from_candid|with|implies|invariant)\b</string>
```
```suggestion
					<string>\b(assert|break|case|continue|default|debug|debug_show|else|if|ignore|in|for|label|null|return|switch|while|loop|try|throw|catch|do|to_candid|from_candid|with|implies|old|invariant)\b</string>
```

aterga reviewed (changes requested):
Added a few minor suggestions.

aterga committed:
Apply suggestions from code review

aterga committed:
Fix typo

rvanasa committed:
Update CI

rvanasa committed:
Drop Node.js 14.x in CI

rvanasa committed:
Deactivate test workflow for now (requires Nix installation)

rvanasa committed:
Explain CI blocker in comment

ggreif committed:
track `motoko` ToT

ggreif committed:
explain how to track the Motoko repo

ggreif committed:
tweak section

by adding more commands

rvanasa committed:
Bump 'vsce' to new package name

rvanasa committed:
Merge branch 'viper' of https://github.com/dfinity/vscode-motoko into viper

rvanasa committed:
Merge

rvanasa committed:
Tentatively re-enable test CI with Nix

rvanasa committed:
Use Cachix action in CI (without auth token)

rvanasa committed:
Deactivate CI (#145)

rvanasa committed:
Bump dependencies

rvanasa committed:
Run formatter; organize imports

rvanasa committed:
Fix 'Restart language server' command

rvanasa committed:
0.0.3

rvanasa committed:
Refactor

rvanasa committed:
Add support for WSL (#198)

* Bump compiler version

* Patch viperTools path for WSL environment

* 0.0.4

* Remove unused import

* 0.1.0

aterga committed:
Draft: Fix minor issues to enable Serokell to use the IDE for local testing (#281)

* first

* Use built-in 'moc.js'

* Reduce diff

* Bump version

* Bump 'motoko' npm package

---------

Co-authored-by: rvanasa <ryan.vandersmith@dfinity.org>
